src/libraries/System.Security.Cryptography/tests/OpenSslNamedKeysTests.manual.cs

```diff
@@ -174,17 +225,202 @@ public static void Engine_UsePublicKey()
+                Assert.NotEqual(data, badSignature);
+                Assert.NotEqual(badSignature, signature);
+                Assert.True(ecdsaPub.VerifyData(data, signature, HashAlgorithmName.SHA256));
+                Assert.False(ecdsaPub.VerifyData(data, badSignature, HashAlgorithmName.SHA256));
+                Assert.False(ecdsaBad.VerifyData(data, signature, HashAlgorithmName.SHA256));
+
+                // TPM key is intended for sign/decrypt only, we could theoretically make verify work without needing to export/import by forcing 'default' provider
+                // for this operation but it's most likely misusage on user part and tpm2 provider intentionally didn't allow it so we will follow this logic.
+                Assert.ThrowsAny<CryptographicException>(() => ecdsaPri.VerifyData(data, signature, HashAlgorithmName.SHA256));
+            }
+
+            // It's TPM so it should not be possible to export parameters
+            Assert.ThrowsAny<CryptographicException>(() => ecdsaPri.ExportParameters(includePrivateParameters: true));
+        }
+
+        [ConditionalFact(nameof(ShouldRunProviderEcDhTests))]
+        public static void Provider_TPM2ECDH()
+        {
+            using SafeEvpPKeyHandle priKeyHandle = SafeEvpPKeyHandle.OpenKeyFromProvider(Tpm2ProviderName, TpmEcDhKeyHandleUri);
+            using ECDiffieHellman alicePri = new ECDiffieHellmanOpenSsl(priKeyHandle);
+            using ECDiffieHellman alicePub = ECDiffieHellman.Create();
+            alicePub.ImportParameters(alicePri.ExportParameters(includePrivateParameters: false));
+
+            using ECDiffieHellman bobPri = ECDiffieHellman.Create();
+            bobPri.KeySize = alicePri.KeySize;
+
+            using ECDiffieHellman bobPub = ECDiffieHellman.Create();
+            bobPub.ImportParameters(bobPri.ExportParameters(includePrivateParameters: false));
+
+            byte[] sharedKeyFromAlice;
+            using (ECDiffieHellmanPublicKey bobPublic = bobPub.PublicKey)
+            {
+                sharedKeyFromAlice = alicePri.DeriveKeyMaterial(bobPublic);
+
+                Assert.NotEmpty(sharedKeyFromAlice);
+
+                byte firstByte = sharedKeyFromAlice[0];
+                bool allSame = sharedKeyFromAlice.All((x) => x == firstByte);
+                Assert.False(allSame, "all bytes of shared key are the same");
+            }
+
+            using (ECDiffieHellmanPublicKey alicePublic = alicePub.PublicKey)
+            {
+                byte[] sharedKeyFromBob = bobPri.DeriveKeyMaterial(alicePublic);
+                Assert.Equal(sharedKeyFromAlice, sharedKeyFromBob);
+            }
+        }
+
+        [ConditionalFact(nameof(ShouldRunProviderRsaSignTests))]
+        [ActiveIssue("https://github.com/tpm2-software/tpm2-openssl/issues/115")]
+        // or a workaround API proposal
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/104080")]
+        public static void Provider_TPM2SignRsa()
```

Comment:
Isn't this only a problem with PSS, but Pkcs1 should work fine?

Reply:
I think they never implemented that in tpm2 but I also couldn't find any issue about it (probably because of Bleicenbacher). I'm getting `error:47000003:tpm2::unknown algorithm` when signing and quickly looking at their code also doesn't seem they handle that (I haven't tried directly with Esys APIs so not sure if that's a hardware or tpm2 or both)